### PR TITLE
[Storage] Support Immutability Policy in AzureRm.Storage

### PR DESCRIPTION
**Pull request by blueww**
## Description

The feature is target to release on 5/18.
The design pass review on 
https://github.com/Azure/azure-powershell-pr/wiki/%5BAzureRM.Storage%5D-ImmutabilityPolicy 

The feature depends on the merged PR https://github.com/Azure/azure-powershell/pull/6068
The change only contains one commit "[Storage] Support Immunablity Policy ", others commits from PR #6068 

Please help to review, but don't merge this PR before the AzureRM.Storage module is released for PR #6068.

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by blueww** [comment]
There's an error in Travis CI build, but I don't understand the meaning .
There's actually no file with name "EnableAzureStorageContainerLease.cs" in my code.

> Blob/EnableAzureStorageContainerLease.cs(56,47): error CS1061: 'IBlobContainersOperations' does not contain a definition for 'Lease' and no extension method 'Lease' accepting a first argument of type 'IBlobContainersOperations' could be found (are you missing a using directive or an assembly reference?) [/home/travis/build/Azure/azure-powershell/src/ResourceManager/Storage/Commands.Management.Storage/Commands.Management.Storage.Netcore.csproj] [/home/travis/build/Azure/azure-powershell/build.proj]

**Comment by blueww** [comment]
And for the StaticAnalysis error, can we just show the error in the console, or give a link to download the error file. Or it will be very difficult to find the error.

**Comment by blueww** [comment]
On Demand Test pass: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/819/

**Comment by praries880** [comment]
@blueww WRT the compilation error,  This seems to be a file you checked in : https://github.com/Azure/azure-powershell/pull/6151/files#diff-11460cb64bf0a95c185f8506c1d9e5

It is part of the PR

Also, I dont see static analysis errors...

**Comment by blueww** [comment]
@praries880 
I have removed the additional file, it's no necessary. Thanks for point it out!

For the static analysis errors, I have resolved it. But it's very inconvenient since can't see the error in the build log, and I need to build again on location machine and see the static analysis errors.

**Comment by praries880** [comment]
@blueww you need to take care of the merge conflicts in the changelog file

**Comment by blueww** [comment]
@praries880 
The conflict is resolved.
If no other issues, we can merge this after the AzureRM.Storage 5.0.1 is released.

**Comment by blueww** [comment]
@praries880  @markcowl 
I have resolved the conflict. Let me know if anything still blocked the merge of the PR.

**Comment by praries880** [comment]
@blueww  Changing target branch to "AzureAzureRM.Storage.Location " as per @markcowl

**Comment by blueww** [comment]
@praries880 

I see the target is changed to "AzureAzureRM.Storage.Location ". Let me know if any thing still blockd the merge.

**Comment by praries880** [comment]
@blueww can you resolve the merge conflict?

**Comment by blueww** [comment]
@praries880 
I have rebase the PR on the "AzureRM.Storage.Location" branch.
All check passed.

Let me know if anything still blocks the merged.

File: src/ResourceManager/Storage/Commands.Management.Storage/Blob/AddAzureStorageContainerLegalHold.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        [ValidateNotNullOrEmpty]
+        public string[] Tag { get; set; }
+        
+        public override void ExecuteCmdlet()
+        {
+            base.ExecuteCmdlet();
+            if (ShouldProcess(this.Name, "Set container LegalHold"))
+            {
+                switch (ParameterSetName)
+                {
```
[the remark this answers is not in the document]
Comment (blueww) [reply src/ResourceManager/Storage/Commands.Management.Storage/Blob/AddAzureStorageContainerLegalHold.cs]: "AccountNameParameterSet" is the default, we don't need to extract the account resourcegroup and name in the parameter set, so just go to default.

File: src/ResourceManager/Storage/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj [hunk 2]
```diff
@@ -175,9 +176,15 @@
     <None Include="ScenarioTests\StorageAccountTests.ps1">
       <CopyToOutputDirectory>Always</CopyToOutputDirectory>
     </None>
+    <None Include="ScenarioTests\StorageBlobTests.ps1">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
```
[the remark this answers is not in the document]
Comment (blueww) [reply src/ResourceManager/Storage/Commands.Management.Storage.Test/Commands.Management.Storage.Test.csproj]: Fixed.

File: src/ResourceManager/Storage/Commands.Management.Storage/Blob/AddAzureStorageContainerLegalHold.cs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+        [Parameter(Mandatory = true,
+            HelpMessage = "Storage container object",
+            ValueFromPipeline = true,
+            ValueFromPipelineByPropertyName = true,
+            ParameterSetName = ContainerObjectParameterSet)]
+        [ValidateNotNullOrEmpty]
+        public PSContainer Container { get; set; }
+
+        [Parameter(HelpMessage = "Container PublicAccess", Mandatory = true)]
```
[the remark this answers is not in the document]
Comment (blueww) [reply src/ResourceManager/Storage/Commands.Management.Storage/Blob/AddAzureStorageContainerLegalHold.cs]: good catch. Fixed.

File: src/ResourceManager/Storage/Commands.Management.Storage/Blob/GetAzureStorageContainer.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            Mandatory = false,
+            ValueFromPipeline = true,
+            ValueFromPipelineByPropertyName = true)]
+        public string Name { get; set; }
+
+        public override void ExecuteCmdlet()
+        {
+            base.ExecuteCmdlet();
+
+            if (ParameterSetName == AccountObjectParameterSet)
```
[the remark this answers is not in the document]
Comment (blueww) [reply src/ResourceManager/Storage/Commands.Management.Storage/Blob/GetAzureStorageContainer.cs]: "AccountNameParameterSet" is the default, for other parameter set, we  need to extract the account resourcegroup and name, but "AccountNameParameterSet" it don't need to do so. So"AccountNameParameterSet" will go to the default and do nothing.

**Comment by blueww** [comment]
@praries880 

I have resolved all comments from you.

For "AccountNameParameterSet", it's the default paramterset, for other parameter sets, we need to extract the account resourcegroup and name from the input object, but "AccountNameParameterSet"  don't need to do so. So"AccountNameParameterSet" will go to the default  in swtich and do nothing. Then the following code will use the ResourceGroupName and account name to execute the cmdlet.

**Comment by blueww** [comment]
@markcowl ， @praries880 
As I talked with Mark today, the feature team would like to have the release on 5/21 ( same day with Oauth).
So would you please help to merge this PR, and give the release candidate together for Oauth and worm (Immutability Policy)?

I will validate it on my Monday, so you can get the result before you Monday.
If every goes smooth, we can release it on next Monday 5/21.

**Comment by praries880** [comment]
Versioning changes at : https://github.com/Azure/azure-powershell/pull/6255

**Comment by praries880** [comment]
Sign job with versioning changes : https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/ps-sign/631/

**Pull request merged**